File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/afb69d139dc9c6d65ecca9bb5333c9e3.java
import java.io.*;
import java.util.*;

class Main {
    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        public FastScanner() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    static class HashedString {
        String s;
        List<Long> MOD, POW;
        List<List<Long>> hash;
        List<List<Long>> pows;

        String substr(int L, int n) {
            return s.substring(L, L + n);
        }

        int size() {
            return s.length();
        }

        char charAt(int n) {
            return s.charAt(n);
        }

        List<Long> getHash(int L, int R) {
            List<Long> ret = new ArrayList<>();
            for (int i = 0; i < hash.size(); ++i) {
                ret.add(nmod(hash.get(i).get(R) - hash.get(i).get(L) * pows.get(i).get(R - L), MOD.get(i)));
            }
            return ret;
        }

        boolean equal(int L1, int R1, int L2, int R2) {
            return getHash(L1, R1).equals(getHash(L2, R2));
        }

        void addChar(char c) {
            s += c;
            for (int i = 0; i < hash.size(); ++i) {
                long m = MOD.get(i), p = POW.get(i), cp = pows.get(i).get(pows.get(i).size() - 1), cm = hash.get(i).get(hash.get(i).size() - 1);
                cm = (cm * p + c) % m;
                cp = cp * p % m;
                hash.get(i).add(cm);
                pows.get(i).add(cp);
            }
        }

        void addString(String oth) {
            for (int i = 0; i < hash.size(); ++i) {
                for (char c : oth.toCharArray()) {
                    addChar(c);
                }
            }
        }

        HashedString(String s, int n) {
            this.s = s;
            MOD = new ArrayList<>();
            POW = new ArrayList<>();
            hash = new ArrayList<>();
            pows = new ArrayList<>();
            for (int i = 0; i < n; ++i) {
                MOD.add(genRandomPrime());
                POW.add(genRandomPrime());
                hash.add(new ArrayList<>());
                pows.add(new ArrayList<>());
            }
            addString(s);
        }
    }

    static int allocator_pos = 0;
    static char[] allocator_memory = new char[(int) FAST_ALLOCATOR_MEMORY];

    static void* operator new(size_t n) {
        char* res = allocator_memory + allocator_pos;
        allocator_pos += n;
        assert(allocator_pos <= (int) FAST_ALLOCATOR_MEMORY);
        return (void*) res;
    }

    static void operator delete(void*) {
    }

    static final int MAXN = 3 * 100000 + 42;
    static int n;

    static class Treap {
        Treap left = null;
        Treap right = null;
        long t;
        int y;
        long prefix_sum;
        long speed;
        long push = 0;
        long min_prefix_sum_in_subtree;

        Treap(long t, long val, long speed) {
            this.t = t;
            this.y = mrand();
            this.prefix_sum = val;
            this.speed = speed;
            this.min_prefix_sum_in_subtree = val;
        }
    }

    static Treap merge(Treap a, Treap b) {
        if (a == null) return b;
        if (b == null) return a;
        if (a.y < b.y) {
            push(a);
            a.right = merge(a.right, b);
            recalc(a);
            return a;
        } else {
            push(b);
            b.left = merge(a, b.left);
            recalc(b);
            return b;
        }
    }

    static void vertex_add(Treap a, long push) {
        a.prefix_sum += push;
        a.min_prefix_sum_in_subtree += push;
        a.push += push;
    }

    static void push(Treap a) {
        long push = a.push;
        if (push == 0) return;
        if (a.left != null) {
            vertex_add(a.left, push);
        }
        if (a.right != null) {
            vertex_add(a.right, push);
        }
        a.push = 0;
    }

    static long smin(Treap a) {
        return a != null ? a.min_prefix_sum_in_subtree : INF;
    }

    static void recalc(Treap a) {
        a.min_prefix_sum_in_subtree = Math.min(Math.min(smin(a.left), smin(a.right)), a.prefix_sum);
    }

    static Treap getLeast(Treap a) {
        push(a);
        if (a.left == null) {
            return a;
        }
        return getLeast(a.left);
    }

    static Treap getGreatest(Treap a) {
        push(a);
        if (a.right == null) {
            return a;
        }
        return getGreatest(a.right);
    }

    static Treap findFirstLess(Treap a, long v) {
        push(a);
        if (smin(a.left) <= v) {
            return findFirstLess(a.left, v);
        }
        if (a.prefix_sum <= v) return a;
        return findFirstLess(a.right, v);
    }

    static void printTree(Treap a) {
        if (a == null) return;
        printTree(a.left);
        System.err.print(a.t + " ");
        printTree(a.right);
    }

    static final long INF = 1_000_000_000_000_000_000L;

    static long nmod(long a, long m) {
        return (a % m + m) % m;
    }

    static int mrand() {
        return Math.abs((1L * rand() << 15) + rand());
    }

    static long genRandomPrime() {
        long L = 1_000_000_000 + 42;
        L += rand() % 4242;
        while (!isPrime(L)) ++L;
        return L;
    }

    static boolean isPrime(long n) {
        for (long i = 2; i * i <= n; ++i)
            if (n % i == 0) return false;
        return true;
    }

    static List<Long> factor(long n) {
        List<Long> ans = new ArrayList<>();
        for (long i = 2; i * i <= n; ++i)
            while (n % i == 0)
                ans.add(i);
        if (n != 1)
            ans.add(n);
        return ans;
    }

    static List<Long> divisors(long n) {
        List<Long> ret = new ArrayList<>();
        for (long i = 1; i * i <= n; ++i) {
            if (n % i == 0) {
                ret.add(i);
                if (i * i != n)
                    ret.add(n / i);
            }
        }
        return ret;
    }

    static List<Pair<Long, Integer>> factorPows(long n) {
        List<Long> fac = factor(n);
        List<Pair<Long, Integer>> ret = new ArrayList<>();
        int L = 0;
        while (L < fac.size()) {
            int R = L;
            while (R < fac.size() && fac.get(R).equals(fac.get(L))) ++R;
            ret.add(new Pair<>(fac.get(L), R - L));
            L = R;
        }
        return ret;
    }

    static void init() {
        srand(System.currentTimeMillis());
    }

    static int readInt(FastScanner scanner) {
        return scanner.nextInt();
    }

    static long readLong(FastScanner scanner) {
        return scanner.nextLong();
    }

    static double readDouble(FastScanner scanner) {
        return scanner.nextDouble();
    }

    static int[] readIntArray(FastScanner scanner, int n) {
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }
        return arr;
    }

    static long[] readLongArray(FastScanner scanner, int n) {
        long[] arr = new long[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextLong();
        }
        return arr;
    }

    static double[] readDoubleArray(FastScanner scanner, int n) {
        double[] arr = new double[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextDouble();
        }
        return arr;
    }

    static void writeInt(PrintWriter writer, int x, char end, int output_len) {
        if (x < 0) {
            writer.print('-');
            x = -x;
        }

        char[] s = new char[24];
        int n = 0;
        while (x > 0 || n == 0) {
            s[n++] = (char) ('0' + x % 10);
            x /= 10;
        }
        while (n < output_len) {
            s[n++] = '0';
        }
        while (n-- > 0) {
            writer.print(s[n]);
        }
        if (end != 0) {
            writer.print(end);
        }
    }

    static void writeWord(PrintWriter writer, String s) {
        writer.print(s);
    }

    static void writeDouble(PrintWriter writer, double x, int output_len) {
        if (x < 0) {
            writer.print('-');
            x = -x;
        }
        int t = (int) x;
        writeInt(writer, t, (char) 0, 0);
        x -= t;
        writer.print('.');
        for (int i = output_len - 1; i > 0; i--) {
            x *= 10;
            t = Math.min(9, (int) x);
            writer.print((char) ('0' + t));
            x -= t;
        }
        x *= 10;
        t = Math.min(9, (int) (x + 0.5));
        writer.print((char) ('0' + t));
    }

    public static void main(String[] args) {
        FastScanner scanner = new FastScanner();
        PrintWriter writer = new PrintWriter(System.out);
        init();
        Treap root = new Treap(0, 0, 0);
        root = merge(root, new Treap(2_000_000_000, 0, 0));
        int q = readInt(scanner);
        int t;
        int s, type;
        for (int i = 0; i < q; ++i) {
            type = readInt(scanner);
            if (type == 1) {
                t = readInt(scanner);
                s = readInt(scanner);
                Pair<Treap, Treap> lr = split(root, t);
                Treap vr = getLeast(lr.s);
                Treap vl = getGreatest(lr.f);
                long psum = vl.prefix_sum + (t - vl.t) * vl.speed;
                long new_psum = psum + (vr.t - t) * s;
                vertex_add(lr.s, new_psum - vr.prefix_sum);
                root = merge(lr.f, merge(new Treap(t, psum, s), lr.s));
            } else if (type == 2) {
                t = readInt(scanner);
                Pair<Treap, Treap> ltr = split(root, t);
                Pair<Treap, Treap> yr = split(ltr.s, t + 1);
                Treap vl = getGreatest(ltr.f);
                Treap vr = getLeast(yr.s);
                long new_psum = vl.prefix_sum + (vr.t - vl.t) * vl.speed;
                vertex_add(yr.s, new_psum - vr.prefix_sum);
                root = merge(ltr.f, yr.s);
            } else {
                int L = readInt(scanner);
                int R = readInt(scanner);
                long v = readInt(scanner);
                if (v == 0) {
                    writer.println(L);
                    continue;
                }
                Pair<Treap, Treap> ltr = split(root, L);
                Pair<Treap, Treap> yr = split(ltr.s, R + 1);
                if (yr.f == null) {
                    writer.println(-1);
                } else {
                    Treap vl = getLeast(yr.f);
                    Treap vr = getGreatest(yr.f);
                    long searching_v = vl.prefix_sum - v;
                    if (yr.f.min_prefix_sum_in_subtree > searching_v) {
                        long new_psum = vr.prefix_sum + (R - vr.t) * vr.speed;
                        if (new_psum > searching_v) {
                            writer.println(-1);
                        } else {
                            writer.println(vr.t + (double) (searching_v - vr.prefix_sum) / vr.speed);
                        }
                    } else {
                        Treap h = findFirstLess(yr.f, searching_v);
                        Pair<Treap, Treap> ylr = split(yr.f, h.t);
                        Treap vyl = getGreatest(ylr.f);
                        writer.println(vyl.t + (double) (searching_v - vyl.prefix_sum) / vyl.speed);
                        yr.f = merge(ylr.f, ylr.s);
                    }
                }
                root = merge(ltr.f, merge(yr.f, yr.s));
            }
        }
        writer.flush();
    }
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/afb69d139dc9c6d65ecca9bb5333c9e3.py
import sys
import random
import time

class HashedString:
    def __init__(self, s, n=1):
        self.s = s
        self.MOD = []
        self.POW = []
        self.hash = []
        self.pows = []
        for i in range(n):
            self.MOD.append(self.gen_random_prime())
            self.POW.append(self.gen_random_prime())
            self.hash.append([0])
            self.pows.append([1])
        self += s
    
    def substr(self, L, n):
        return self.s[L:L+n]
    
    def size(self):
        return len(self.s)
    
    def __getitem__(self, n):
        return self.s[n]
    
    def get_hash(self, L, R):
        ret = []
        for i in range(len(self.hash)):
            ret.append(self.nmod(self.hash[i][R] - self.hash[i][L] * self.pows[i][R - L], self.MOD[i]))
        return ret
    
    def equal(self, L1, R1, L2, R2):
        return self.get_hash(L1, R1) == self.get_hash(L2, R2)
    
    def __iadd__(self, oth):
        for i in range(len(self.hash)):
            for c in oth:
                self.s += c
                m = self.MOD[i]
                p = self.POW[i]
                cp = self.pows[i][-1]
                cm = self.hash[i][-1]
                cm = (cm * p + ord(c)) % m
                cp = cp * p % m
                self.hash[i].append(cm)
                self.pows[i].append(cp)
        return self
    
    def gen_random_prime(self, L=1e9):
        L += random.randint(0, 4242)
        while not self.is_prime(L):
            L += 1
        return L
    
    def is_prime(self, n):
        for i in range(2, int(n ** 0.5) + 1):
            if n % i == 0:
                return False
        return True
    
    def factor(self, n):
        ans = []
        for i in range(2, int(n ** 0.5) + 1):
            while n % i == 0:
                ans.append(i)
                n //= i
        if n != 1:
            ans.append(n)
        return ans
    
    def divisors(self, n):
        ret = []
        for i in range(1, int(n ** 0.5) + 1):
            if n % i == 0:
                ret.append(i)
                if i * i != n:
                    ret.append(n // i)
        return ret
    
    def factor_pows(self, n):
        fac = self.factor(n)
        ret = []
        L = 0
        while L < len(fac):
            R = L
            while R < len(fac) and fac[R] == fac[L]:
                R += 1
            ret.append((fac[L], R - L))
            L = R
        return ret
    
    def nmod(self, a, m):
        return (a % m + m) % m
    
    def mrand(self):
        return abs((1 * random.randint(0, (1 << 15))) + random.randint(0, (1 << 15)))

class Treap:
    def __init__(self, t, val, speed):
        self.left = None
        self.right = None
        self.t = t
        self.y = self.mrand()
        self.prefix_sum = val
        self.speed = speed
        self.push = 0
        self.min_prefix_sum_in_subtree = val
    
    def vertex_add(self, push):
        self.prefix_sum += push
        self.min_prefix_sum_in_subtree += push
        self.push += push
    
    def push(self):
        push = self.push
        if not push:
            return
        if self.left:
            self.left.vertex_add(push)
        if self.right:
            self.right.vertex_add(push)
        self.push = 0
    
    def smin(self):
        return self.min_prefix_sum_in_subtree if self else INF
    
    def recalc(self):
        self.min_prefix_sum_in_subtree = min(self.smin(self.left), self.smin(self.right), self.prefix_sum)
    
    def merge(self, a, b):
        if not a:
            return b
        if not b:
            return a
        if a.y < b.y:
            a.push()
            a.right = self.merge(a.right, b)
            a.recalc()
            return a
        else:
            b.push()
            b.left = self.merge(a, b.left)
            b.recalc()
            return b
    
    def split(self, a, k):
        if not a:
            return (a, a)
        a.push()
        if a.t < k:
            l, r = self.split(a.right, k)
            a.right = l
            a.recalc()
            return (a, r)
        else:
            l, r = self.split(a.left, k)
            a.left = r
            a.recalc()
            return (l, a)
    
    def get_least(self, a):
        a.push()
        if not a.left:
            return a
        return self.get_least(a.left)
    
    def get_greatest(self, a):
        a.push()
        if not a.right:
            return a
        return self.get_greatest(a.right)
    
    def find_first_less(self, a, v):
        a.push()
        if self.smin(a.left) <= v:
            return self.find_first_less(a.left, v)
        if a.prefix_sum <= v:
            return a
        return self.find_first_less(a.right, v)
    
    def print_tree(self, a):
        if not a:
            return
        self.print_tree(a.left)
        print(a.t, end=' ')
        self.print_tree(a.right)

def readString():
    return sys.stdin.readline().strip()

def readInt():
    return int(readString())

def readDouble():
    return float(readString())

def writeInt(x, end=' ', output_len=-1):
    sys.stdout.write(str(x))
    sys.stdout.write(end)

def writeWord(s):
    sys.stdout.write(s)

def writeDouble(x, output_len=0):
    sys.stdout.write(format(x, '.{}f'.format(output_len)))

def flush():
    sys.stdout.flush()

def isEof():
    return peekChar() == -1

def getChar():
    global buf_pos, buf_len
    if buf_pos == buf_len:
        buf_pos = 0
        buf_len = sys.stdin.buffer.readinto(buf)
        if buf_pos == buf_len:
            return -1
    return buf[buf_pos]

def peekChar():
    global buf_pos, buf_len
    if buf_pos == buf_len:
        buf_pos = 0
        buf_len = sys.stdin.buffer.readinto(buf)
        if buf_pos == buf_len:
            return -1
    return buf[buf_pos]

def seekEof():
    while peekChar() != -1 and peekChar() <= 32:
        buf_pos += 1
    return peekChar() == -1

def skipBlanks():
    while not isEof() and buf[buf_pos] <= 32:
        buf_pos += 1

def readChar():
    c = getChar()
    while c != -1 and c <= 32:
        c = getChar()
    return c

def init():
    random.seed(time.time())
    sys.stdin = open('input.txt', 'r')
    sys.stdout = open('output.txt', 'w')

def main():
    init()
    root = Treap(0, 0, 0)
    root = root.merge(root, Treap(2e9, 0, 0))
    q = readInt()
    for _ in range(q):
        type = readInt()
        if type == 1:
            t = readInt()
            s = readInt()
            l, r = root.split(root, t)
            vr = root.get_least(r)
            vl = root.get_greatest(l)
            psum = vl.prefix_sum + (t - vl.t) * vl.speed
            new_psum = psum + (vr.t - t) * s
            vr.vertex_add(new_psum - vr.prefix_sum)
            root = root.merge(l, root.merge(Treap(t, psum, s), r))
        elif type == 2:
            t = readInt()
            l, tr = root.split(root, t)
            y, r = root.split(tr, t + 1)
            vl = root.get_greatest(l)
            vr = root.get_least(r)
            new_psum = vl.prefix_sum + (vr.t - vl.t) * vl.speed
            vr.vertex_add(new_psum - vr.prefix_sum)
            root = root.merge(l, r)
        else:
            L = readInt()
            R = readInt()
            v = readInt()
            if v == 0:
                writeInt(L, '\n')
                continue
            l, yr = root.split(root, L)
            y, r = root.split(yr, R + 1)
            if not y:
                writeInt(-1, '\n')
            else:
                vl = root.get_least(y)
                vr = root.get_greatest(y)
                searching_v = vl.prefix_sum - v
                if y.min_prefix_sum_in_subtree > searching_v:
                    new_psum = vr.prefix_sum + (R - vr.t) * vr.speed
                    if new_psum > searching_v:
                        writeInt(-1, '\n')
                    else:
                        writeDouble(vr.t + (searching_v - vr.prefix_sum) / vr.speed, '\n')
                else:
                    h = root.find_first_less(y, searching_v)
                    yl, yr = root.split(y, h.t)
                    vyl = root.get_greatest(yl)
                    writeDouble(vyl.t + (searching_v - vyl.prefix_sum) / vyl.speed, '\n')
                    y = root.merge(yl, yr)
            root = root.merge(l, root.merge(y, r))
    flush()

if __name__ == "__main__":
    main()
